In [43]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb

Load core datasets

In [44]:
base_path = "../data/raw/california/"

incidents = pd.read_csv(base_path + "NIBRS_incident.csv")
victims = pd.read_csv(base_path + "NIBRS_VICTIM.csv")
offenders = pd.read_csv(base_path + "NIBRS_OFFENDER.csv")
offenses = pd.read_csv(base_path + "NIBRS_OFFENSE.csv")
arrestees = pd.read_csv(base_path + "NIBRS_ARRESTEE.csv")
agencies = pd.read_csv(base_path + "agencies.csv", encoding="cp1252")

Load lookup tables

In [45]:
weapon_type = pd.read_csv(base_path + "NIBRS_WEAPON_TYPE.csv")
race = pd.read_csv(base_path + "REF_RACE.csv")
ethnicity = pd.read_csv(base_path + "NIBRS_ETHNICITY.csv")
age = pd.read_csv(base_path + "NIBRS_AGE.csv")
location_type = pd.read_csv(base_path + "NIBRS_LOCATION_TYPE.csv")
relationship = pd.read_csv(base_path + "NIBRS_RELATIONSHIP.csv")
state = pd.read_csv(base_path + "REF_STATE.csv")
offense_type = pd.read_csv(base_path + "NIBRS_OFFENSE_TYPE.csv")


Secondary tables

In [46]:
victim_offender_rel = pd.read_csv(base_path + "NIBRS_VICTIM_OFFENDER_REL.csv")
victim_injury = pd.read_csv(base_path + "NIBRS_VICTIM_INJURY.csv")
victim_circumstances = pd.read_csv(base_path + "NIBRS_VICTIM_CIRCUMSTANCES.csv")
weapon = pd.read_csv(base_path + "NIBRS_WEAPON.csv")
nibrs_property = pd.read_csv(base_path + "NIBRS_PROPERTY.csv")

Verification

In [47]:
print("incidents:", incidents.shape)
print("victims:", victims.shape)
print("offenders:", offenders.shape)
print("arrestees:", arrestees.shape)
print("agencies:", agencies.shape)
print("offenses:", offenses.shape)

incidents: (1188704, 15)
victims: (1362826, 16)
offenders: (1351022, 11)
arrestees: (370032, 18)
agencies: (655, 59)
offenses: (1369571, 8)


## Data Inspection Stage

### Incidents Dataset

In [52]:
print("Head: ")
incidents.head()

Head: 


,data_year,agency_id,incident_id,nibrs_month_id,cargo_theft_flag,submission_date,incident_date,report_date_flag,incident_hour,cleared_except_id,cleared_except_date,incident_status,data_home,orig_format,did
0,2024,937,190226279,44856163,f,2024-05-02 20:04:52.306,2024-03-05,f,14.0,6,NaN,ACCEPTED,NaN,X,246196887
1,2024,937,185618101,44732720,f,2024-02-08 04:53:04.531,2024-01-01,f,5.0,6,NaN,ACCEPTED,NaN,X,231242773
2,2024,937,187487274,44732720,f,2024-03-11 01:20:17.253,2024-01-01,f,5.0,6,NaN,WARNINGS,NaN,X,237181432
3,2024,937,185618105,44732720,f,2024-02-08 04:53:05.521,2024-01-01,f,8.0,6,NaN,ACCEPTED,NaN,X,231242785
4,2024,937,185618107,44732720,f,2024-02-08 04:53:06.053,2024-01-01,f,10.0,6,NaN,ACCEPTED,NaN,X,231242791


In [53]:
print("Shape: ")
incidents.shape

Shape: 


(1188704, 15)

In [54]:
print("Info: ")
incidents.info()

Info: 
<class 'pandas.DataFrame'>
RangeIndex: 1188704 entries, 0 to 1188703
Data columns (total 15 columns):
 #   Column               Non-Null Count    Dtype  
---  ------               --------------    -----  
 0   data_year            1188704 non-null  int64  
 1   agency_id            1188704 non-null  int64  
 2   incident_id          1188704 non-null  int64  
 3   nibrs_month_id       1188704 non-null  int64  
 4   cargo_theft_flag     1188704 non-null  str    
 5   submission_date      1188704 non-null  str    
 6   incident_date        1188704 non-null  str    
 7   report_date_flag     1188704 non-null  str    
 8   incident_hour        1161668 non-null  float64
 9   cleared_except_id    1188704 non-null  int64  
 10  cleared_except_date  9972 non-null     str    
 11  incident_status      1188704 non-null  str    
 12  data_home            0 non-null        float64
 13  orig_format          1188704 non-null  str    
 14  did                  1188704 non-null  int64  
dtypes:

In [55]:
print("Describe: ")
incidents.describe()

Describe: 


,data_year,agency_id,incident_id,nibrs_month_id,incident_hour,cleared_except_id,data_home,did
count,1188704.0,1.188704e+06,1.188704e+06,1.188704e+06,1.161668e+06,1.188704e+06,0.0,1.188704e+06
mean,2024.0,2.005169e+03,1.972396e+08,4.493933e+07,1.270263e+01,5.972773e+00,NaN,2.688154e+08
std,0.0,4.468076e+03,5.924535e+06,1.028431e+05,6.688144e+00,3.105171e-01,NaN,1.903355e+07
min,2024.0,9.370000e+02,1.831516e+08,4.468028e+07,0.000000e+00,1.000000e+00,NaN,2.235317e+08
25%,2024.0,1.205000e+03,1.924747e+08,4.489581e+07,8.000000e+00,6.000000e+00,NaN,2.537818e+08
50%,2024.0,1.387000e+03,1.975485e+08,4.495744e+07,1.400000e+01,6.000000e+00,NaN,2.699755e+08
75%,2024.0,1.582000e+03,2.020021e+08,4.501594e+07,1.800000e+01,6.000000e+00,NaN,2.841711e+08
max,2024.0,4.288200e+04,2.084446e+08,4.516236e+07,2.300000e+01,6.000000e+00,NaN,3.043163e+08


In [56]:
print("Sum of null : ")
incidents.isnull().sum()

Sum of null : 


data_year                    0
agency_id                    0
incident_id                  0
nibrs_month_id               0
cargo_theft_flag             0
submission_date              0
incident_date                0
report_date_flag             0
incident_hour            27036
cleared_except_id            0
cleared_except_date    1178732
incident_status              0
data_home              1188704
orig_format                  0
did                          0
dtype: int64

#### Incidents Observations

- Grain of dataset: Each row represents one crime incident reported in California (2024). No aggregation is done at row level.

- Key identifier: incident_id (used to uniquely identify each incident and link with other tables)

- Time-related columns:
  - submission_date → when the record was submitted (string, needs datetime conversion)
  - incident_date → when the crime happened (string, needs datetime conversion)
  - incident_hour → hour of incident (partially missing ~27,036 values)
  - nibrs_month_id → encoded month identifier
  - cleared_except_date → date of case clearance (very sparse)

- Missing values:
  - incident_hour → missing for some records (unknown time)
  - cleared_except_date → mostly missing (cases not cleared or not updated)
  - data_home → completely empty (no usable information)

- Column types issue:
  - Dates are stored as strings (not datetime yet)
  - Most IDs are integers (used for linking, not calculation)

- System/unused fields:
  - data_home → empty column (no data present)
  - orig_format → data source format indicator (not used for analysis)
  - did → internal system identifier (not needed for analysis)

### Victims Dataset

In [57]:
print("Head: ")
victims.head()

Head: 


,data_year,victim_id,incident_id,victim_seq_num,victim_type_id,assignment_type_id,activity_type_id,outside_agency_id,age_id,age_num,sex_code,race_id,ethnicity_id,resident_status_code,age_range_low_num,age_code_range_high
0,2024,211571444,190226279,1,8,NaN,NaN,NaN,104,NS,X,99,50,NaN,NaN,NaN
1,2024,206775798,186056807,1,1,NaN,NaN,NaN,104,NS,X,99,50,NaN,NaN,NaN
2,2024,210441858,189249900,1,3,NaN,NaN,NaN,104,NS,X,99,50,NaN,NaN,NaN
3,2024,208514605,187573271,1,4,NaN,NaN,NaN,52,49,F,10,10,R,49.0,NaN
4,2024,206775799,186056808,1,4,NaN,NaN,NaN,45,42,M,10,10,N,42.0,NaN


In [58]:
print("Shape: ")
victims.shape

Shape: 


(1362826, 16)

In [59]:
print("Info: ")
victims.info()

Info: 
<class 'pandas.DataFrame'>
RangeIndex: 1362826 entries, 0 to 1362825
Data columns (total 16 columns):
 #   Column                Non-Null Count    Dtype  
---  ------                --------------    -----  
 0   data_year             1362826 non-null  int64  
 1   victim_id             1362826 non-null  int64  
 2   incident_id           1362826 non-null  int64  
 3   victim_seq_num        1362826 non-null  int64  
 4   victim_type_id        1362826 non-null  int64  
 5   assignment_type_id    11754 non-null    float64
 6   activity_type_id      11754 non-null    float64
 7   outside_agency_id     309 non-null      float64
 8   age_id                1362826 non-null  int64  
 9   age_num               1362826 non-null  str    
 10  sex_code              1362826 non-null  str    
 11  race_id               1362826 non-null  int64  
 12  ethnicity_id          1362826 non-null  int64  
 13  resident_status_code  671677 non-null   str    
 14  age_range_low_num     904884 non-null 

In [61]:
print("Describe: ")
victims.describe()

Describe: 


,data_year,victim_id,incident_id,victim_seq_num,victim_type_id,assignment_type_id,activity_type_id,outside_agency_id,age_id,race_id,ethnicity_id,age_range_low_num,age_code_range_high
count,1362826.0,1.362826e+06,1.362826e+06,1.362826e+06,1.362826e+06,11754.000000,11754.000000,309.000000,1.362826e+06,1.362826e+06,1.362826e+06,904884.000000,15747.000000
mean,2024.0,2.196991e+08,1.973018e+08,1.294640e+00,4.035353e+00,2.866258,5.531479,9530.595469,6.368222e+01,4.703564e+01,3.273290e+01,40.102957,68.210453
std,0.0,6.804472e+06,5.923663e+06,2.185109e+00,1.902258e+00,1.861625,3.527756,14469.115160,3.190030e+01,4.168768e+01,1.693403e+01,17.511834,26.322703
min,2024.0,2.034494e+08,1.831516e+08,1.000000e+00,1.000000e+00,1.000000,1.000000,937.000000,1.000000e+00,1.000000e+01,1.000000e+01,1.000000,2.000000
25%,2024.0,2.141572e+08,1.924801e+08,1.000000e+00,4.000000e+00,2.000000,2.000000,1340.000000,3.600000e+01,1.000000e+01,2.000000e+01,27.000000,45.000000
50%,2024.0,2.200305e+08,1.975829e+08,1.000000e+00,4.000000e+00,2.000000,6.000000,1577.000000,5.600000e+01,2.000000e+01,4.000000e+01,38.000000,70.000000
75%,2024.0,2.252783e+08,2.021480e+08,1.000000e+00,4.000000e+00,3.000000,8.000000,1752.000000,1.040000e+02,9.900000e+01,5.000000e+01,52.000000,98.000000
max,2024.0,2.324754e+08,2.084446e+08,1.820000e+02,9.000000e+00,7.000000,11.000000,35795.000000,1.040000e+02,9.900000e+01,5.000000e+01,98.000000,99.000000


In [62]:
print("Sum of null : ")
victims.isnull().sum()

Sum of null : 


data_year                     0
victim_id                     0
incident_id                   0
victim_seq_num                0
victim_type_id                0
assignment_type_id      1351072
activity_type_id        1351072
outside_agency_id       1362517
age_id                        0
age_num                       0
sex_code                      0
race_id                       0
ethnicity_id                  0
resident_status_code     691149
age_range_low_num        457942
age_code_range_high     1347079
dtype: int64

#### Victims Observations

- Grain of dataset: Each row represents one victim record linked to a specific incident in 2024. One incident can have multiple victim records.

- Key relationship: incident_id (used to connect victim data with the main incidents table)

- Multiple victims per incident: Yes. This is supported by repeated incident_id values and victim_seq_num indicating multiple victims within the same incident.

- Important attributes:
  - victim_type_id (type/category of victim)
  - age_num (victim age, stored as string)
  - sex_code (gender of victim)
  - race_id, ethnicity_id (demographic information)
  - resident_status_code (residence status, partially missing)

- Missing patterns:
  - Very high missing values in assignment_type_id, activity_type_id, and outside_agency_id (mostly unused in this dataset)
  - resident_status_code is missing for many records (~50%)
  - age_range_low_num partially missing
  - age_code_range_high mostly missing

- Use:
  - Study victim demographics (age, sex, race, ethnicity)
  - Analyze victim distribution across incidents
  - Compare victim characteristics across crime types (after merging with incidents)

### Offenders Dataset

In [65]:
print("Head: ")
offenders.head()

Head: 


,data_year,offender_id,incident_id,offender_seq_num,age_id,age_num,sex_code,race_id,ethnicity_id,age_range_low_num,age_range_high_num
0,2024,216455996,190226279,1,22,19,M,10,10,19.0,NaN
1,2024,211204264,185618101,1,36,33,M,10,10,33.0,NaN
2,2024,213331030,187487274,1,22,19,M,20,20,19.0,NaN
3,2024,213331031,187487274,2,103,00,U,98,40,NaN,NaN
4,2024,211204270,185618105,2,37,34,F,20,20,34.0,NaN


In [66]:
print("Shape: ")
offenders.shape

Shape: 


(1351022, 11)

In [67]:
print("Info: ")
offenders.info()

Info: 
<class 'pandas.DataFrame'>
RangeIndex: 1351022 entries, 0 to 1351021
Data columns (total 11 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   data_year           1351022 non-null  int64  
 1   offender_id         1351022 non-null  int64  
 2   incident_id         1351022 non-null  int64  
 3   offender_seq_num    1351022 non-null  int64  
 4   age_id              1351022 non-null  int64  
 5   age_num             1351022 non-null  str    
 6   sex_code            1351022 non-null  str    
 7   race_id             1351022 non-null  int64  
 8   ethnicity_id        1351022 non-null  int64  
 9   age_range_low_num   736732 non-null   float64
 10  age_range_high_num  85553 non-null    float64
dtypes: float64(2), int64(7), str(2)
memory usage: 113.4 MB


In [68]:
print("Describe: ")
offenders.describe()

Describe: 


,data_year,offender_id,incident_id,offender_seq_num,age_id,race_id,ethnicity_id,age_range_low_num,age_range_high_num
count,1351022.0,1.351022e+06,1.351022e+06,1.351022e+06,1.351022e+06,1.351022e+06,1.351022e+06,736732.000000,85553.000000
mean,2024.0,2.245824e+08,1.973209e+08,8.508574e-01,6.792714e+01,5.177287e+01,3.675001e+01,34.254183,39.475319
std,0.0,6.774778e+06,5.916696e+06,1.028017e+00,3.406560e+01,4.268158e+01,1.635394e+01,13.023971,17.733219
min,2024.0,2.083963e+08,1.831516e+08,0.000000e+00,4.000000e+00,1.000000e+01,1.000000e+01,1.000000,3.000000
25%,2024.0,2.191448e+08,1.925785e+08,0.000000e+00,3.500000e+01,1.000000e+01,2.000000e+01,25.000000,29.000000
50%,2024.0,2.248816e+08,1.975847e+08,1.000000e+00,5.800000e+01,2.000000e+01,5.000000e+01,33.000000,35.000000
75%,2024.0,2.301140e+08,2.021486e+08,1.000000e+00,1.040000e+02,9.900000e+01,5.000000e+01,42.000000,49.000000
max,2024.0,2.373167e+08,2.084446e+08,9.900000e+01,1.040000e+02,9.900000e+01,5.000000e+01,98.000000,99.000000


In [69]:
print("Sum of null : ")
offenders.isnull().sum()

Sum of null : 


data_year                   0
offender_id                 0
incident_id                 0
offender_seq_num            0
age_id                      0
age_num                     0
sex_code                    0
race_id                     0
ethnicity_id                0
age_range_low_num      614290
age_range_high_num    1265469
dtype: int64

#### Offenders Observations

- Grain: Each row represents one offender involved in a crime incident. One incident can have multiple offenders.

- Relationship to incidents: Linked using incident_id. This is a many-to-one relationship (many offenders per incident).

- Key fields:
  - offender_id → unique identifier for each offender record
  - incident_id → links offender to incident dataset
  - offender_seq_num → order of offenders within the same incident
  - age_num → offender age (string format, more reliable than range fields)
  - sex_code, race_id, ethnicity_id → demographic attributes

- Missing data patterns:
  - age_range_low_num is missing for ~614,290 records
  - age_range_high_num is missing for ~1,265,469 records
  - age range fields are highly incomplete and not reliable for most records
  - core identifiers and demographic fields are fully populated

- Use:
  - Study offender demographics (age, sex, race, ethnicity)
  - Analyze number of offenders per incident
  - Compare offender profiles across different incident types (after merging with incidents dataset)

###  Offenses Dataset

In [70]:
print("Head: ")
offenses.head()

Head: 


,data_year,offense_id,incident_id,offense_code,attempt_complete_flag,location_id,num_premises_entered,method_entry_code
0,2024,226036771,190226279,35A,C,28,NaN,NaN
1,2024,220713974,185618101,520,C,2,NaN,NaN
2,2024,222869305,187487274,240,A,25,NaN,NaN
3,2024,222869306,187487274,520,C,25,NaN,NaN
4,2024,220713979,185618105,13A,C,35,NaN,NaN


In [71]:
print("Shape: ")
offenses.shape

Shape: 


(1369571, 8)

In [75]:
print("Info: ")
offenses.info()

Info: 
<class 'pandas.DataFrame'>
RangeIndex: 1369571 entries, 0 to 1369570
Data columns (total 8 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   data_year              1369571 non-null  int64  
 1   offense_id             1369571 non-null  int64  
 2   incident_id            1369571 non-null  int64  
 3   offense_code           1369571 non-null  str    
 4   attempt_complete_flag  1369571 non-null  str    
 5   location_id            1369571 non-null  int64  
 6   num_premises_entered   6072 non-null     float64
 7   method_entry_code      77145 non-null    str    
dtypes: float64(1), int64(4), str(3)
memory usage: 83.6 MB


In [73]:
print("Describe: ")
offenses.describe()

Describe: 


,data_year,offense_id,incident_id,location_id,num_premises_entered
count,1369571.0,1.369571e+06,1.369571e+06,1.369571e+06,6072.000000
mean,2024.0,2.341390e+08,1.972775e+08,3.367958e+01,1.934453
std,0.0,6.804472e+06,5.923833e+06,1.989600e+01,6.415472
min,2024.0,2.178822e+08,1.831516e+08,1.000000e+00,1.000000
25%,2024.0,2.286281e+08,1.924786e+08,2.500000e+01,1.000000
50%,2024.0,2.344831e+08,1.975601e+08,3.300000e+01,1.000000
75%,2024.0,2.396206e+08,2.020485e+08,3.500000e+01,1.000000
max,2024.0,2.469539e+08,2.084446e+08,9.800000e+01,99.000000


In [74]:
print("Sum of null : ")
offenses.isnull().sum()

Sum of null : 


data_year                      0
offense_id                     0
incident_id                    0
offense_code                   0
attempt_complete_flag          0
location_id                    0
num_premises_entered     1363499
method_entry_code        1292426
dtype: int64

#### Offenses Observations

- Grain: Each row represents a single offense recorded within an incident. One incident can have multiple offenses.

- Relationship to incidents: Linked using incident_id (many offenses per incident).

- Key fields:
  - offense_id → unique identifier for each offense record
  - incident_id → links offense to incident dataset
  - offense_code → type of crime committed
  - attempt_complete_flag → whether offense was attempted or completed
  - location_id → coded location type of offense

- Missing data patterns:
  - num_premises_entered is mostly missing (~1.36M missing) and only applies to specific offenses like burglary
  - method_entry_code is also highly missing (~1.29M missing), indicating it is only relevant for certain offense types

- Use:
  - Identify most common crime types (using offense_code)
  - Compare attempted vs completed offenses
  - Analyze crime distribution across location types
  - Study patterns of specific crime methods (when data is available)

### Arrestees Dataset

In [76]:
print("Head: ")
arrestees.head()

Head: 


,data_year,arrestee_id,incident_id,arrestee_seq_num,arrest_date,arrest_type_id,multiple_indicator,offense_code,age_id,age_num,sex_code,race_id,ethnicity_id,resident_code,under_18_disposition_code,clearance_ind,age_range_low_num,age_range_high_num
0,2024,56488775,185618101,1,2024-01-01,1,N,520,36,33,M,10,10,U,NaN,NaN,33.0,NaN
1,2024,57055028,187487274,1,2024-01-01,1,N,240,22,19,M,20,40,U,NaN,NaN,19.0,NaN
2,2024,56488777,185618105,1,2024-01-01,1,N,13A,42,39,M,20,20,U,NaN,NaN,39.0,NaN
3,2024,57055029,187487276,1,2024-01-02,1,N,13A,27,24,M,10,40,U,NaN,NaN,24.0,NaN
4,2024,56488780,185618129,1,2024-01-02,1,N,13A,44,41,M,10,40,U,NaN,NaN,41.0,NaN


In [77]:
print("Shape: ")
arrestees.shape

Shape: 


(370032, 18)

In [78]:
print("Info: ")
arrestees.info()

Info: 
<class 'pandas.DataFrame'>
RangeIndex: 370032 entries, 0 to 370031
Data columns (total 18 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   data_year                  370032 non-null  int64  
 1   arrestee_id                370032 non-null  int64  
 2   incident_id                370032 non-null  int64  
 3   arrestee_seq_num           370032 non-null  int64  
 4   arrest_date                370032 non-null  str    
 5   arrest_type_id             370032 non-null  int64  
 6   multiple_indicator         370032 non-null  str    
 7   offense_code               370032 non-null  str    
 8   age_id                     370032 non-null  int64  
 9   age_num                    370032 non-null  int64  
 10  sex_code                   370032 non-null  str    
 11  race_id                    370032 non-null  int64  
 12  ethnicity_id               370032 non-null  int64  
 13  resident_code              294648

In [79]:
print("Describe: ")
arrestees.describe()

Describe: 


,data_year,arrestee_id,incident_id,arrestee_seq_num,arrest_type_id,age_id,age_num,race_id,ethnicity_id,clearance_ind,age_range_low_num,age_range_high_num
count,370032.0,3.700320e+05,3.700320e+05,370032.000000,370032.000000,370032.000000,370032.000000,370032.000000,370032.000000,0.0,370030.000000,96.000000
mean,2024.0,6.032356e+07,1.975316e+08,1.152811,1.544493,38.968938,35.968668,15.280803,20.706506,NaN,35.968094,21.114583
std,0.0,1.902060e+06,5.881445e+06,1.070348,0.771455,12.311276,12.310968,15.070134,12.425237,NaN,12.311055,11.820473
min,2024.0,5.572697e+07,1.831516e+08,1.000000,1.000000,10.000000,0.000000,10.000000,10.000000,NaN,7.000000,14.000000
25%,2024.0,5.889256e+07,1.930604e+08,1.000000,1.000000,30.000000,27.000000,10.000000,10.000000,NaN,27.000000,17.000000
50%,2024.0,6.048026e+07,1.979850e+08,1.000000,1.000000,38.000000,35.000000,10.000000,20.000000,NaN,35.000000,17.000000
75%,2024.0,6.192716e+07,2.024424e+08,1.000000,2.000000,46.000000,43.000000,10.000000,20.000000,NaN,43.000000,17.000000
max,2024.0,6.387485e+07,2.084445e+08,99.000000,3.000000,103.000000,99.000000,98.000000,50.000000,NaN,93.000000,81.000000


In [80]:
print("Sum of null : ")
arrestees.isnull().sum()

Sum of null : 


data_year                         0
arrestee_id                       0
incident_id                       0
arrestee_seq_num                  0
arrest_date                       0
arrest_type_id                    0
multiple_indicator                0
offense_code                      0
age_id                            0
age_num                           0
sex_code                          0
race_id                           0
ethnicity_id                      0
resident_code                 75384
under_18_disposition_code    349200
clearance_ind                370032
age_range_low_num                 2
age_range_high_num           369936
dtype: int64

#### Arrestees Observations

- Grain: Each row represents one arrested individual linked to a crime incident.

- Relationship to incidents: Linked using incident_id (many arrestees per incident possible, but most incidents have only one arrestee).

- Key fields:
  - arrestee_id → unique identifier for arrest record
  - incident_id → links to incident dataset
  - arrestee_seq_num → order of arrestees within an incident
  - arrest_date → date of arrest
  - arrest_type_id → type/category of arrest
  - offense_code → offense linked to the arrest
  - age_num, sex_code, race_id, ethnicity_id → demographic details

- Missing data patterns:
  - resident_code missing for ~75,000 records
  - under_18_disposition_code mostly missing (~349,000 records)
  - age_range_high_num almost entirely missing (~369,936 records)
  - clearance_ind completely empty in this dataset

- Use:
  - Study arrest rates by offense type
  - Analyze demographics of arrested individuals
  - Compare arrests vs incidents to understand enforcement patterns
  - Examine juvenile vs adult arrest distribution (limited due to missing data)

### Agencies Dataset

In [82]:
print("Head: ")
agencies.head()

Head: 


,yearly_agency_id,agency_id,data_year,ori,legacy_ori,covered_by_legacy_ori,direct_contributor_flag,dormant_flag,dormant_year,reporting_type,...,nibrs_leoka_start_date,nibrs_ct_start_date,nibrs_multi_bias_start_date,nibrs_off_eth_start_date,covered_flag,county_name,msa_name,publishable_flag,participated,nibrs_participated
0,9372024,937,2024,CA0010000,CA0010000,NaN,N,N,NaN,I,...,2022-10-01,2022-10-01,2022-10-01,2022-10-01,N,ALAMEDA,"San Francisco-Oakland-Fremont, CA",Y,Y,Y
1,9382024,938,2024,CA0010100,CA0010100,NaN,N,N,NaN,I,...,2022-07-01,2022-07-01,2022-07-01,2022-07-01,N,ALAMEDA,"San Francisco-Oakland-Fremont, CA",Y,Y,Y
2,9392024,939,2024,CA0010200,CA0010200,NaN,N,N,NaN,I,...,2022-07-01,2022-07-01,2022-07-01,2022-07-01,N,ALAMEDA,"San Francisco-Oakland-Fremont, CA",Y,Y,Y
3,9402024,940,2024,CA0010300,CA0010300,NaN,N,N,NaN,I,...,2023-12-01,2023-12-01,2023-12-01,2023-12-01,N,ALAMEDA,"San Francisco-Oakland-Fremont, CA",Y,Y,Y
4,9412024,941,2024,CA0010400,CA0010400,NaN,N,N,NaN,I,...,2021-05-01,2021-05-01,2021-05-01,2021-05-01,N,ALAMEDA,"San Francisco-Oakland-Fremont, CA",Y,Y,Y


In [83]:
print("Shape: ")
agencies.shape

Shape: 


(655, 59)

In [84]:
print("Info: ")
agencies.info()

Info: 
<class 'pandas.DataFrame'>
RangeIndex: 655 entries, 0 to 654
Data columns (total 59 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   yearly_agency_id                655 non-null    int64  
 1   agency_id                       655 non-null    int64  
 2   data_year                       655 non-null    int64  
 3   ori                             655 non-null    str    
 4   legacy_ori                      655 non-null    str    
 5   covered_by_legacy_ori           1 non-null      str    
 6   direct_contributor_flag         655 non-null    str    
 7   dormant_flag                    655 non-null    str    
 8   dormant_year                    0 non-null      float64
 9   reporting_type                  655 non-null    str    
 10  ucr_agency_name                 655 non-null    str    
 11  ncic_agency_name                460 non-null    str    
 12  pub_agency_name                 655 non-

In [85]:
print("Describe: ")
agencies.describe()

Describe: 


,yearly_agency_id,agency_id,data_year,dormant_year,state_id,division_code,region_code,population,submitting_agency_id,population_group_id,...,pop_sort_order,summary_rape_def,male_officer,male_civilian,male_officer+male_civilian,female_officer,female_civilian,female_officer+female_civilian,officer_rate,employee_rate
count,6.550000e+02,655.000000,655.0,0.0,655.0,655.0,655.0,6.550000e+02,655.000000,655.000000,...,655.000000,0.0,413.000000,413.000000,413.000000,413.000000,413.000000,413.000000,611.000000,611.000000
mean,1.134980e+08,11349.596947,2024.0,NaN,6.0,9.0,4.0,4.580894e+04,23431.572519,10.688550,...,10.688550,NaN,116.675545,25.210654,141.886199,18.331719,43.796610,62.128329,1.361538,1.954010
std,1.524905e+08,15249.050493,0.0,NaN,0.0,0.0,0.0,1.744676e+05,854.899177,4.210527,...,4.210527,NaN,502.369692,111.812502,607.619906,90.808017,144.203175,223.998527,8.838274,11.675981
min,9.372024e+06,937.000000,2024.0,NaN,6.0,9.0,4.0,0.000000e+00,23357.000000,3.000000,...,3.000000,NaN,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1.289702e+07,1289.500000,2024.0,NaN,6.0,9.0,4.0,0.000000e+00,23357.000000,8.000000,...,8.000000,NaN,16.000000,1.000000,18.000000,2.000000,6.000000,9.000000,0.000000,0.000000
50%,1.585202e+07,1585.000000,2024.0,NaN,6.0,9.0,4.0,7.910000e+03,23357.000000,10.000000,...,10.000000,NaN,35.000000,4.000000,43.000000,5.000000,15.000000,22.000000,0.900000,1.300000
75%,3.387520e+08,33875.000000,2024.0,NaN,6.0,9.0,4.0,4.901850e+04,23357.000000,11.000000,...,11.000000,NaN,79.000000,12.000000,96.000000,12.000000,33.000000,45.000000,1.300000,1.900000
max,4.288220e+08,42882.000000,2024.0,NaN,6.0,9.0,4.0,3.796352e+06,34995.000000,22.000000,...,22.000000,NaN,7159.000000,1557.000000,8270.000000,1661.000000,1879.000000,3165.000000,214.600000,282.900000


In [86]:
print("Sum of null : ")
agencies.isnull().sum()

Sum of null : 


yearly_agency_id                    0
agency_id                           0
data_year                           0
ori                                 0
legacy_ori                          0
covered_by_legacy_ori             654
direct_contributor_flag             0
dormant_flag                        0
dormant_year                      655
reporting_type                      0
ucr_agency_name                     0
ncic_agency_name                  195
pub_agency_name                     0
pub_agency_unit                   429
agency_status                       0
state_id                            0
state_name                          0
state_abbr                          0
state_postal_abbr                   0
division_code                       0
division_name                       0
region_code                         0
region_name                         0
region_desc                         0
agency_type_name                    0
population                          0
submitting_a

#### Agencies Observations

- Grain: Each row represents one law enforcement agency for the year 2024.

- Key field: agency_id (used to link with the Incidents dataset).

- Useful attributes:
  - ucr_agency_name / pub_agency_name → agency name
  - agency_type_name → type of law enforcement agency
  - agency_status → agency’s operational status
  - county_name, msa_name, state_name → agency location
  - population → population served by the agency
  - reporting_type, participated, nibrs_participated → reporting and participation information

- Missing data patterns:
  - covered_by_legacy_ori is almost completely empty.
  - dormant_year is completely empty.
  - pub_agency_unit and some staffing-related columns contain many missing values.
  - Most important identification and location columns have no missing values.

- Use:
  - Join with the Incidents dataset using agency_id.
  - Identify which agency handled each incident.
  - Compare crime across different agencies, counties, or metropolitan areas.
  - Add agency information (name, location, population, etc.) to incident-level analysis.

### Lookup Tables Overview

In [94]:
print("Weapon type:")
print("Shape:", weapon_type.shape)
display(weapon_type.head())


print("\n\nRace:")
print("Shape:", race.shape)
display(race.head())


print("\n\nEthnicity:")
print("Shape:", ethnicity.shape)
display(ethnicity.head())


print("\n\nAge:")
print("Shape:", age.shape)
display(age.head())


print("\n\nLocation type:")
print("Shape:", location_type.shape)
display(location_type.head())


print("\n\nRelationship:")
print("Shape:", relationship.shape)
display(relationship.head())


print("\n\nState:")
print("Shape:", state.shape)
display(state.head())


print("\n\nOffense type:")
print("Shape:", offense_type.shape)
display(offense_type.head())

Weapon type:
Shape: (38, 4)


,weapon_id,weapon_code,weapon_name,shr_flag
0,21,20,Knife/Cutting Instrument,t
1,31,30,Blunt Object,t
2,32,35,Motor Vehicle/Vessel,f
3,33,50,Poison,t
4,34,60,Explosives,t




Race:
Shape: (12, 7)


,race_id,race_code,race_desc,sort_order,start_year,end_year,notes
0,98,U,Unknown,98,NaN,NaN,Race is explicitly unknown
1,10,W,White,10,NaN,NaN,NaN
2,20,B,Black or African American,20,NaN,NaN,NaN
3,30,I,American Indian or Alaska Native,30,NaN,NaN,NaN
4,40,A,Asian,40,2013.0,NaN,Includes Asian Indian




Ethnicity:
Shape: (5, 3)


,ethnicity_id,ethnicity_code,ethnicity_name
0,10,H,Hispanic or Latino
1,20,N,Not Hispanic or Latino
2,30,M,Multiple
3,40,U,Unknown
4,50,X,Not Specified




Age:
Shape: (104, 3)


,age_id,age_code,age_name
0,1,NN,Under 24 Hours
1,2,NB,1-6 Days Old
2,3,BB,7-364 Days Old
3,103,00,Unknown
4,104,NS,Not Specified




Location type:
Shape: (47, 3)


,location_id,location_code,location_name
0,1,37,Abandoned/Condemned Structure
1,2,1,Air/Bus/Train Terminal
2,3,38,Amusement Park
3,4,39,Arena/Stadium/Fairgrounds/Coliseum
4,5,40,ATM Separate from Bank




Relationship:
Shape: (30, 3)


,relationship_id,relationship_code,relationship_name
0,1,AQ,Victim Was Acquaintance
1,2,BE,Victim Was Babysittee
2,3,BG,Victim Was Boyfriend/Girlfriend
3,4,CF,Victim Was Child of Boyfriend or Girlfriend
4,5,CH,Victim Was Child




State:
Shape: (59, 9)


,state_id,division_id,state_name,state_code,state_abbr,state_postal_abbr,state_fips_code,state_pub_freq_months,change_user
0,1,10,Alaska,50,AK,AK,2.0,12,NaN
1,2,7,Alabama,1,AL,AL,1.0,12,NaN
2,3,8,Arkansas,3,AR,AR,5.0,12,NaN
3,4,1,American Samoa,54,AS,AS,60.0,12,NaN
4,5,9,Arizona,2,AZ,AZ,4.0,12,NaN




Offense type:
Shape: (86, 8)


,offense_code,offense_name,crime_against,ct_flag,hc_flag,hc_code,offense_category_name,offense_group
0,09C,Justifiable Homicide,Person,f,f,,Homicide Offenses,A
1,09A,Murder and Nonnegligent Manslaughter,Person,f,t,01,Homicide Offenses,A
2,09B,Negligent Manslaughter,Person,f,t,,Homicide Offenses,A
3,11A,Rape,Person,f,t,02,Sex Offenses,A
4,11B,Sodomy,Person,f,t,02,Sex Offenses,A


#### Lookup Tables Observations

- Purpose: These tables act as decoding dictionaries that convert coded values (IDs) into readable labels such as names, categories, or descriptions.

- Role in dataset: They are not used for direct statistical analysis but are essential for interpreting other datasets like incidents, victims, offenders, and arrests.

- Structure: Each table maps an ID column to a meaningful label.

- Tables included:
  - Weapon type
  - Race
  - Ethnicity
  - Age
  - Location type
  - Relationship
  - State
  - Offense type

- Use: Convert numeric or coded fields into human-readable form

## Data Ingestion Summary
- Incidents dataset is the central table for analysis.
- Victims, offenders, and arrestees link to incidents using incident_id (one-to-many relationship).
- Agencies link to incidents using agency_id.
- Offenses provide details of crimes within each incident (multiple per incident).
- Lookup tables are used to decode coded fields, not for direct analysis.
- Many columns contain missing values and will need cleaning before analysis.
- Overall dataset is relational and requires joins across multiple tables for analysis.

## Data Cleaning Stage

### Data integrity check

In [96]:
print("uniqueness test: ")
incidents["incident_id"].is_unique

uniqueness test: 


True

In [97]:
print("\n\nduplicate count: ")
incidents["incident_id"].duplicated().sum()



duplicate count: 


np.int64(0)

In [98]:
print("\n\ninspect actual duplicates: ")
incidents[incidents["incident_id"].duplicated(keep=False)].sort_values("incident_id")



inspect actual duplicates: 


,data_year,agency_id,incident_id,nibrs_month_id,cargo_theft_flag,submission_date,incident_date,report_date_flag,incident_hour,cleared_except_id,cleared_except_date,incident_status,data_home,orig_format,did


In [99]:
print("\n\ncheck full-row duplicates: ")
incidents.duplicated().sum()



check full-row duplicates: 


np.int64(0)

In [100]:
print("\n\nisolate true bad duplicates: ")
incidents[incidents.duplicated(keep=False)]



isolate true bad duplicates: 


,data_year,agency_id,incident_id,nibrs_month_id,cargo_theft_flag,submission_date,incident_date,report_date_flag,incident_hour,cleared_except_id,cleared_except_date,incident_status,data_home,orig_format,did


In [101]:
print("\n\ncount incidents per incident_id: ")
incidents["incident_id"].value_counts().describe()



count incidents per incident_id: 


count    1188704.0
mean           1.0
std            0.0
min            1.0
25%            1.0
50%            1.0
75%            1.0
max            1.0
Name: count, dtype: float64

In [102]:
print("\n\nfind worst offenders (if any): ")
incidents["incident_id"].value_counts().head()



find worst offenders (if any): 


incident_id
190226279    1
185618101    1
187487274    1
185618105    1
185618107    1
Name: count, dtype: int64

In [103]:
print("\n\nidentical rows test: ")
incidents[incidents["incident_id"].duplicated(keep=False)]



identical rows test: 


,data_year,agency_id,incident_id,nibrs_month_id,cargo_theft_flag,submission_date,incident_date,report_date_flag,incident_hour,cleared_except_id,cleared_except_date,incident_status,data_home,orig_format,did
